### PR TITLE
finish minimal shell standardization efforts

### LINKED ISSUE #1141: [CI] [proposal] low-efforts to standardize scripting
I have been enjoying learning more shell by patching things here and there in our `bin/` and `_test/` files.
As an outsider, I can see little things that tell the story of these scripts being written over time by many folks.

I propose we standardize a small set of guidelines.
I appreciate Google's [scripting guidelines](https://google.github.io/styleguide/shellguide.html) but feel we probably don't need that degree of specificity now. 

# wishlist
- underscores or dashes for file names
- whether the filename should include `.sh` or not
- executable bit set
- favor a single shell interpreter
    - I'd vote BASH simply because it gives you a tad more expressiveness and seems more plausible. Actual usage here is split between `sh` and `bash`. And we can standardize `set -e`.

### PR DESCRIPTION
**Pull request by efx**
This should be enough to close #1141. 
I
- added extra details from the issue to maintaining
- added a filename linter for anything in `bin/` and `test/` that ends with `.sh`
- added that linter to CI so this PR should fail until I fix those :)

### CODE REVIEW

File: bin/lint_tool_file_names.sh [hunk 1]
```diff
@@ -0,0 +1,17 @@
+#!/usr/bin/env bash
+
+# Prohibit internal shell script file names snake_case
```
Comment (coriolinus) [inline bin/lint_tool_file_names.sh:3]:
```suggestion
# Require that internal shell script file names use snake_case
```

File: bin/lint_tool_file_names.sh [hunk 2]
```diff
@@ -0,0 +1,17 @@
...
+set -eo pipefail
+
+# find a list of files whose names do not match our convention
+errant_files=$(find bin/ _test/ -iname '*\.sh' -exec basename {} \; | grep '[^a-z_.].*')
```
Comment (coriolinus) [inline bin/lint_tool_file_names.sh:7]: I don't entirely understand the regex's intent here, but it looks like what it's actually checking is that there is a single character not in `[a-z_.]`, followed by any number of arbitrary other characters. That feels like an odd requirement to check. I think that if we're just detecting the presence of a forbidden character, we could simplify like

```suggestion
errant_files=$(find bin/ _test/ -iname '*\.sh' -exec basename {} \; | grep '[^a-z_.]')
```

File: bin/lint_tool_file_names.sh [hunk 3]
```diff
@@ -0,0 +1,17 @@
...
+
+if [ -n "$errant_files" ]; then
+    echo "These file names do not follow our snake case convention:"
+    # find them again to print the whole relative path
+    while IFS= read -r file_name; do
+        find . -iname "$file_name"
```
Comment (coriolinus) [inline bin/lint_tool_file_names.sh:13]:
```suggestion
        find . -name "$file_name"
```

In this case we can be pretty confident that we have the named file's casing correct, and can avoid the case-conversion expense.

**Review by coriolinus — APPROVED** [review]
I think this is good to merge once CI is green.

File: bin/lint_tool_file_names.sh (hunk 2 above)
Comment (efx) [reply bin/lint_tool_file_names.sh:7]: > That feels like an odd requirement to check.

Perfect correction to my regular expression naïveté!
I was trying to find any file name that has forbidden characters. So your fix makes it so!

**Pull request merged**